api.go

```diff
@@ -200,7 +200,157 @@ func (r *Router) RouterCallImpl(ctx context.Context,
 	}
 }
 
-// todo: router_map_callrw
+func (r *Router) callStorageUnref(futures map[uuid.UUID]*tarantool.Future, refID int64) {
+	req := tarantool.NewCallRequest("vshard.storage._call")
+	req = req.Args([]interface{}{"storage_unref", refID})
+
+	for id, future := range futures {
+		future.SetError(nil)
+
+		conn := r.idToReplicaset[id].conn
+
+		future = conn.Do(req, pool.RW)
+		future.SetError(nil)
+	}
+}
+
+// RouterMapCallRWImpl perform call function on all masters in the cluster
+// with a guarantee that in case of success it was executed with all
+// buckets being accessible for reads and writes.
+func (r *Router) RouterMapCallRWImpl(
+	ctx context.Context,
+	fnc string,
+	args interface{},
+	opts CallOpts,
+) (map[uuid.UUID]interface{}, error) {
+	if opts.Timeout == 0 {
+		opts.Timeout = CallTimeoutMin
+	}
+
+	timeout := opts.Timeout
+	timeStart := time.Now()
+
+	refID := r.refID.Load()
+	r.refID.Add(1)
+
+	futures := make(map[uuid.UUID]*tarantool.Future, 0)
+
+	defer r.callStorageUnref(futures, refID)
+
+	req := tarantool.NewCallRequest("vshard.storage._call")
+	req = req.Context(ctx)
+
+	// ref stage: send
+
+	req = req.Args([]interface{}{
+		"storage_ref",
+		refID,
+		timeout,
+	})
+
+	for id, replicaset := range r.idToReplicaset {
+		conn := replicaset.conn
+
+		future := conn.Do(req, pool.RW)
+		futures[id] = future
+	}
+
+	// ref stage collect
+
+	totalBucketCount := uint16(0)
+
+	for _, future := range futures {
+		respData, err := future.Get()
```

Comment:
I don't like the synchronous way of getting futures. Maybe we will use list of channels of futures and will call reflect.Select on them?